osbuild/util/fscache.py

```diff
@@ -0,0 +1,918 @@
+    a `*.lock`-file locks the directory it is in, while a lock on any other
+    file just locks that file (unfortunately, we cannot acquire write-locks on
+    directories directly, since it would require opening them for writing,
+    which is not possible on linux). `cache.lock` can be used to guard the
+    entire cache. A write-lock will keep any other parallel operation out,
+    while a read-lock merely acquires cache access (you are still allowed to
+    modify the cache, but need fine-grained locking). Hence, a write-lock on the
+    global `cache.lock` file is only required for operations that cannot use
+    fine-grained locking. The latter requires individual locking for each file
+    or each object store entry you modify. In all those cases you must ensure
+    for parallel modifications, since lock acquisition on file-systems can only
+    be done after opening a file.
+    """
+
+    class MissError(Exception):
+        """Cache Miss Exception
+
+        This error is raised when a cache entry is not found. Due to the
+        shared nature of the cache, a caller must be aware that any entry can
+        be created or deleted by other concurrent operations, at any point in
+        time. Hence, a cache miss only reflects the state of the cache at a
+        particular time under a particular lock.
+        """
+
+    # static parameters
+    _dirname_data = "data"
+    _dirname_objects = "objects"
+    _dirname_stage = "stage"
+    _filename_cache_info = "cache.info"
+    _filename_cache_lock = "cache.lock"
+    _filename_cache_size = "cache.size"
+    _filename_object_info = "object.info"
+    _filename_object_lock = "object.lock"
+
+    # constant properties
+    _appid: Any = None
+    _libc: Any = None
+    _path_cache: Any = None
+
+    # context-manager properties
+    _active: bool = False
+    _bootid: Any = None
+    _lock: Any = None
+    _info: Any = None
+
+    def __init__(self, appid: str, path_cache: str):
+        """Create File System Cache
+
+        This creates a new file-system cache. It does not create the cache, nor
+        access any of its content. You must enter its context-manager to prepare
+        the cache for access. Any access outside of a context-manager will raise
+        an assertion error, unless explicitly stated otherwise.
+
+        Parameters:
+        -----------
+        appid
+            The application-ID of the caller. This can be any random string. It
+            is used to initialize the application-specific boot-ID used to tag
+            caches and detect whether an entry was created during the same boot.
+        path_cache
+            The path to the cache directory. The directory (and the path to it)
+            is created if it does not exist.
+        """
+
+        self._appid = appid
+        self._libc = linux.Libc.default()
+        self._path_cache = path_cache
+
+    @staticmethod
+    def _copy_a(path_from: str, path_to: str):
```

Comment:
We have a bunch of copies of this; even though it is very short, I wonder if it is not something we should extract into `utils`. The advantage would be that this is a prime candidate to move to rust, so that we don't have to selinux label the `cp` binary in the build root.